template/tree_v4.go

```diff
@@ -51,18 +51,18 @@ func (t *TreeV4) CountTags() int {
 	return ret
 }
 
-// add a tag to the node at the input index, storing it in the first position if 'replaceFirst' is true
+// add a tag to the node at the input index, replacing the first position if updateFunc is provided
 // - if matchFunc is non-nil, will enforce uniqueness at this node
 // - returns whether the tag count was increased
-func (t *TreeV4) addTag(tag GeneratedType, nodeIndex uint, matchFunc MatchesFunc, replaceFirst bool) bool {
+func (t *TreeV4) addTag(tag GeneratedType, nodeIndex uint, matchFunc MatchesFunc, updateFunc UpdatesFunc) bool {
 	ret := true
-	if replaceFirst {
+	if updateFunc != nil {
 		if t.nodes[nodeIndex].TagCount == 0 {
 			t.nodes[nodeIndex].TagCount = 1
 		} else {
 			ret = false
 		}
-		t.tags[(uint64(nodeIndex) << 32)] = tag
+		t.tags[(uint64(nodeIndex) << 32)] = updateFunc(t.tags[(uint64(nodeIndex) << 32)])
```

Comment:
This part is incorrect as the existing tag could be random when TagCount was 0. That's one reason I need to work a bit more on this PR.